ios_service: Initiale commit to create the module `ios_service`

earendilfr opened the pull request:
##### SUMMARY
Add a new module ios_service to configure the service part of Cisco IOS devices

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
- ios_service: Configure the services on Cisco IOS and IOS-XE devices
  Has been tested again v16.9.6 of IOS-XE and v15.x of IOS

##### ADDITIONAL INFORMATION
The current stat of the configuration is done again the command `show running-config all` to be sure to have the status of services that are enabled by default on Cisco devices

KB-perByte commented:
@earendilfr Thank you for your contribution, the currently supported version of the collection is Cisco IOS XE 17.6, If you need I can help you out to test the module with the desired version.
Please note, you would need to add proper Unit test cases that have good coverage of the attributes present.
And some integration test cases.
I would help you with the review before we can merge it.
Let me know if you need anything
Regards

KB-perByte commented on plugins/modules/ios_service.py:
```diff
@@ -0,0 +1,425 @@
+short_description: Resource module to configure service.
+description:
+  - This module configures and manages service attributes on IOS platforms
+version_added: 1.0.0
```
This would be the latest cisco.ios version

earendilfr replied:
I have put the current version (`4.5.0`) but should I put `4.6.0` or a tag like `latest` ?

KB-perByte commented on plugins/modules/ios_service.py:
```diff
@@ -0,0 +1,425 @@
+    type: str
+"""
+EXAMPLES = """
+# Using state: merged
```
we are standardizing the way we write out examples, the long-term plan being to generate tests and other components on the basis of examples. 
Please follow the structure as per [ref](https://github.com/ansible-collections/cisco.ios/blob/cc093bba4c2269f84528fb45ea5c58ea66df31f8/plugins/modules/ios_interfaces.py#L123) to add examples.

KB-perByte commented on plugins/modules/ios_service.py:
```diff
@@ -0,0 +1,425 @@
+# Before state:
+# -------------
+
+# switch-ios#sh run | sec ^service|^no service
```
the before state should be strict as per the command used in facts to generate the configuration for the module.

earendilfr replied:
OK: I have updated the comment (and the command use to retrieve the fact for the module is `show running-config all | section ^service ` to be sure to be able to remove some service started by default on some IOS version (service that is recommended to stop by security)

KB-perByte commented on plugins/modules/ios_service.py:
```diff
@@ -0,0 +1,425 @@
+    """
+    required_if = [
+        ("state", "merged", ("config",)),
+        ("state", "replaced", ("config",)),
```
Do we need replaced state for this module?
Does it add any significant use case other than overridden?

earendilfr replied:
Because the `replaced` and `overridden` action have the same result, I have keep only the state `replaced`

KB-perByte commented on plugins/module_utils/network/ios/config/service/service.py:
```diff
@@ -0,0 +1,161 @@
+            )
+
+        # if state us replaced or overridden, force the default value in wantd
+        elif self.state == "replaced" or self.state == "overridden":
```
Both states seems to have similar behaviour

earendilfr replied:
As previous comment: only `replaced` has been retained

KB-perByte commented:
@earendilfr I hope the module was initially scaffolded with https://github.com/ansible-community/ansible.content_builder
Which takes care of most of the recent practices I pointed out in the review.
Please consider using it.
Regards

earendilfr commented on changelogs/fragments/ios_service_changes.yml:
```diff
@@ -0,0 +1,3 @@
+---
+doc_changes:
```
OK Changed as indiquated

earendilfr commented on plugins/modules/ios_service.py:
```diff
@@ -0,0 +1,425 @@
+# service password-encryption
+
+"""
+RETURN = """
```
I have updated with your comment and the source file indiquated as example

earendilfr commented:
> @earendilfr I hope the module was initially scaffolded with https://github.com/ansible-community/ansible.content_builder Which takes care of most of the recent practices I pointed out in the review. Please consider using it. Regards

Thanks a lot for this information: I didn't have seen that tool to generate the initial document for the module.
One question: did I need also to perform a pull request to the repo [resource_module_models](https://github.com/ansible-network/resource_module_models) to add the YAML file to generate this module ?

KB-perByte commented:
@earendilfr resource_modules_model is just a repo where we store our models, you can definitely push the model there and tag me for review. you may refer to this [guide](https://github.com/ansible-community/ansible.content_builder/blob/main/docs/rm_dev_guide.md) for resource module development.
Let me know if you need any help.
Please consider tagging me in replies, helps me get back with a reply faster :) 
Regards

earendilfr commented:
> @earendilfr resource_modules_model is just a repo where we store our models, you can definitely push the model there and tag me for review. you may refer to this [guide](https://github.com/ansible-community/ansible.content_builder/blob/main/docs/rm_dev_guide.md) for resource module development. Let me know if you need any help. Please consider tagging me in replies, helps me get back with a reply faster :) Regards

@KB-perByte : perfect, I have do the PR for the models

KB-perByte commented on plugins/module_utils/network/ios/facts/service/service.py:
```diff
@@ -0,0 +1,80 @@
+    def get_service_data(self, connection):
+        return connection.get("show running-config all | section ^service ")
+
+    def sort_list_dicts(self, objs):
```
Sorting is not required to generate facts, a lot of our modules have it we are planning a cleanup activity.
Regards

KB-perByte commented on plugins/module_utils/network/ios/rm_templates/service.py:
```diff
@@ -0,0 +1,457 @@
+    # fmt: off
+    PARSERS = [
+        {
+            "name": "call_home",
```
Neat Parsers tbh ++

KB-perByte commented on plugins/module_utils/network/ios/config/service/service.py:
```diff
@@ -0,0 +1,156 @@
+            "telnet_zeroidle",
+            "unsupported_transceiver",
+        ]
+        self.list_parsers = [
```
there is only one parser, here a declared variable can be skipped

KB-perByte commented on plugins/module_utils/network/ios/config/service/service.py:
```diff
@@ -0,0 +1,156 @@
+
+    def _compare_lists_attrs(self, want, have):
+        """Compare list of dict"""
+        for _parser in self.list_parsers:
```
list iteration is not needed as there is just 1 parser, if there are more list parsers added we can add a list and iterate.
This is a cosmetic comment, I am fine if you like how it already is.

earendilfr replied:
As previous comment

KB-perByte commented on plugins/modules/ios_service.py:
```diff
@@ -0,0 +1,689 @@
+short_description: Resource module to configure service.
+description:
+  - This module configures and manages service attributes on IOS platforms
+version_added: 4.5.0
```
```suggestion
version_added: 4.6.0
```

KB-perByte commented on plugins/modules/ios_service.py:
```diff
@@ -0,0 +1,689 @@
+# - service timestamps log datetime msec localtime show-timezone year
+# - service timestamps debug datetime
+
+# Use parsed
```
```suggestion
# Using parsed
```

KB-perByte commented on plugins/modules/ios_service.py:
```diff
@@ -0,0 +1,689 @@
+  returned: when I(state) is C(rendered)
+  type: list
+  sample:
+    - sample command 1
```
and here

KB-perByte commented on plugins/modules/ios_service.py:
```diff
@@ -0,0 +1,689 @@
+        required_if=[
+            ["state", "merged", ["config"]],
+            ["state", "replaced", ["config"]],
+            ["state", "overridden", ["config"]],
```
overridden is not a supported state right?

earendilfr replied:
Oups... 😅

KB-perByte commented:
Hey @earendilfr 
Thanks again for a super clean module contribution, loved the parsers. I added some more comments as part of the review.
I plan to merge this PR soon.

earendilfr commented:
> Hey @earendilfr Thanks again for a super clean module contribution, loved the parsers. I added some more comments as part of the review. I plan to merge this PR soon.

@KB-perByte : you're welcome.

I have learned a lot of think concerning how to write a `ios` modules on this PR.
I think it's OK to your last remarks

KB-perByte commented:
@earendilfr do you plan to add unit tests/integration tests for the module, I see it would take down the overall coverage of the collection.
Integration test can be considered optional but unit tests are a must
Regards

earendilfr commented:
> @earendilfr do you plan to add unit tests/integration tests for the module, I see it would take down the overall coverage of the collection. Integration test can be considered optional but unit tests are a must Regards

I have added today the unit test (I think).
And so, I have fixed some bugs in weird case...
